File: toy_model_project/toy_model.py
import torch
import time

# Define a toy linear model
class Net(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super(Net, self).__init__()
        self.linear = torch.nn.Linear(in_features, out_features)

    def forward(self, x):
        return self.linear(x)

if __name__ == '__main__':
    # Start measuring the total execution time
    start_total = time.time()

    # Measure time to initialize CUDA context
    start_cuda = time.time()
    device = torch.device("cuda")
    torch.cuda.synchronize()  # Ensure any prior CUDA work is complete
    end_cuda = time.time()
    diff_cuda = end_cuda - start_cuda
    print(f"Time to initialize CUDA context: {diff_cuda:.6f} s")

    # Create an instance of the toy linear model
    net = Net(5, 2)  # 5 input features, 2 output features

    # Move the model to GPU
    net.to(device)
   



    # Create a toy input tensor of size [10, 5] and move it to GPU
    input_tensor = torch.randn(10, 5).to(device)

    # Warm-up: Run inference multiple times
    warmup_runs = 10
    for _ in range(warmup_runs):
        dummy_output = net(input_tensor)

    # Measure time for inference
    start_inference = time.time()
    output = net(input_tensor)
    torch.cuda.synchronize()  # Ensure the inference operation is complete
    end_inference = time.time()
    diff_inference = end_inference - start_inference
    print(f"Time for inference: {diff_inference:.6f} s")

    # End of total execution time measurement
    end_total = time.time()
    diff_total = end_total - start_total
    print(f"Total execution time: {diff_total:.6f} s")

File: toy_model_project/toy_model.cpp
#include <torch/torch.h>
#include <iostream>
#include <chrono>

// Define a toy linear model
struct Net : torch::nn::Module {
    Net(int in_features, int out_features) 
        : linear(register_module("linear", torch::nn::Linear(in_features, out_features))) 
    {}

    torch::Tensor forward(torch::Tensor x) {
        return linear->forward(x);
    }

    torch::nn::Linear linear{nullptr};
};

int main() {
    // Start measuring the total execution time
    auto start_total = std::chrono::high_resolution_clock::now();

    // Measure time to initialize CUDA context
    auto start_cuda = std::chrono::high_resolution_clock::now();
    torch::Device device(torch::kCUDA);
    torch::cuda::synchronize();  // Add synchronization after CUDA context initialization
    auto end_cuda = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff_cuda = end_cuda - start_cuda;
    std::cout << "Time to initialize CUDA context: " << diff_cuda.count() << " s" << std::endl;

    // Create an instance of the toy linear model
    Net net(5, 2);  // 5 input features, 2 output features

    // Move the model to GPU
    net.to(device);

    // Create a toy input tensor of size [10, 5] and move it to GPU
    torch::Tensor input = torch::randn({10, 5}).to(device);

    // Warm-up: Run inference multiple times
    const int warmup_runs = 10;  // you can adjust this number as needed
    for (int i = 0; i < warmup_runs; ++i) {
        torch::Tensor dummy_output = net.forward(input);
    }

    torch::cuda::synchronize();  // Synchronize before starting the timer for inference

    // Measure time for inference
    auto start_inference = std::chrono::high_resolution_clock::now();
    torch::Tensor output = net.forward(input);
    torch::cuda::synchronize();  // Synchronize after inference to get accurate timing
    auto end_inference = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff_inference = end_inference - start_inference;
    std::cout << "Time for inference: " << diff_inference.count() << " s" << std::endl;

    // End of total execution time measurement
    auto end_total = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff_total = end_total - start_total;
    std::cout << "Total execution time: " << diff_total.count() << " s" << std::endl;

    return 0;
}
